### PR TITLE
RFC: Rust dependency versions

### PR DESCRIPTION
**Pull request by jdisanti**
## Motivation and Context
This RFC intends to establish how the smithy-rs project selects Rust dependency versions.

[Rendered](https://github.com/awslabs/smithy-rs/blob/jdisanti-rfc-dependency-versions/design/src/rfcs/rfc0021_dependency_versions.md)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: design/src/rfcs/rfc0021_dependency_versions.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+When in conflict, the security goal takes priority over the compatibility goal.
+
+Categorization of Crates
+------------------------
+
+The Rust crates within smithy-rs can be divided up into two categories:
+
+1. **Library Crates:** Crates that are published to crates.io with the intention that other projects
+   will depend on them via their `Cargo.toml` files. This category does NOT include binaries that
+   are published to crates.io with the intention of being installed with `cargo install`.
```
Comment (david-perez) [inline design/src/rfcs/rfc0021_dependency_versions.md:23]: We don't have any of these do we?

File: design/src/rfcs/rfc0021_dependency_versions.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+2. **Application Crates:** All examples, binaries, tools, standalone tests, or other crates that are
+   not published to crates.io with the intent of being depended on by other projects.
+
+### Code generated crates
+
+Generated clients must be considered library crates even if they're not published since they are intended
+to be pulled into other Rust projects with other dependencies.
+
+Generated servers must be considered application crates.
```
Comment (david-perez) [inline design/src/rfcs/rfc0021_dependency_versions.md:32]: Why? I don't understand the distinction. The sentence above for generated clients applies for generated servers too.

File: design/src/rfcs/rfc0021_dependency_versions.md (hunk 1 above)
Comment (jdisanti) [reply design/src/rfcs/rfc0021_dependency_versions.md:23]: Right now, there's just the `cargo-check-external-types` tool, but I'm hoping to break that out into a separate GItHub repository sometime soon.

File: design/src/rfcs/rfc0021_dependency_versions.md (hunk 2 above)
Comment (jdisanti) [reply design/src/rfcs/rfc0021_dependency_versions.md:32]: Yeah, actually, good point. The server runtime and code generated crates need to be broadly compatible with the libraries used by the server implementers consuming them. I'll fix this.

**Review by david-perez — APPROVED** [review]

**Review by Velfi — APPROVED** [review]

**Pull request merged**